atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresDdlTable.java

```diff
@@ -118,6 +123,48 @@ public void checkDatabaseVersion() {
 
     @Override
     public void compactInternally() {
+        if (compactionSemaphore.tryAcquire()) {
+            try {
+                if (shouldRunCompaction()) {
+                    runCompactOnTable();
+                }
+            } finally {
+                compactionSemaphore.release();
+            }
+        }
+    }
+
+    @VisibleForTesting
+    boolean shouldRunCompaction() {
+        long compactIntervalMillis = config.compactInterval().toMilliseconds();
+        return compactIntervalMillis <= 0 || getMillisSinceLastCompact() >= compactIntervalMillis;
+    }
+
+    /**
+     * Returns the number of milliseconds since the last compaction, or Long.MAX_VALUE if
+     * compaction has never run.
+     */
+    private long getMillisSinceLastCompact() {
+        AgnosticResultSet rs = conns.get().selectResultSetUnregisteredQuery(
+                "SELECT FLOOR(EXTRACT(EPOCH FROM GREATEST( "
+                        + "  last_vacuum, last_autovacuum, last_analyze, last_autoanalyze"
+                        + "))*1000) AS last, "
+                        + "FLOOR(EXTRACT(EPOCH FROM CURRENT_TIMESTAMP)*1000) AS current "
+                        + "FROM pg_stat_user_tables WHERE relname = ?",
+                prefixedTableName());
+
+        AgnosticResultRow row = Iterables.getOnlyElement(rs.rows());
+
+        // last could be null if vacuum has never run
+        long last = row.getLong("last", -1);
+        if (last == -1) {
```

Comment:
nit: any reason we can't use `row.getLongObject` and do a null test here? There is a perf cost to pay in terms of boxing but I don't think is hit that often

Reply:
If we're worried about the overhead of unboxing a single long then we shouldn't be using our SQL library (AgnosticResultSet) but instead use raw PreparedStatements etc.

Reply:
This is a very infrequent query, run by the background sweeper at the end of a sweep, I would argue even an extra DB round trip shouldnt be a concern, but is definitely better. :) Lets not focus on perf here.